### ISSUE TITLE
Xs in alignment, Error codes in "Finished aligning and formatting gene-set steps"

### ISSUE THREAD

**Issue by gkatahon** [AstrobioMike/GToTree#53]
First of all, great tool, works smoothly!

I actually have 3 questions:

1. I noticed that GToTree adds 5 Xs in the final alignment between the different aligned markers. Is there a reason for this?
2. When running the same input data (or different input data sets), I noticed that the order of the marker genes in Partitions.txt is not always the same. Sometimes it seems to be alphabetical, whereas sometimes it's more random (although the random one usually starts with the FbpA marker). Although this doesn't infuence the result, I was just wondering what influences this order.
3. During the _Finished aligning and formatting gene-set_ step, some markers (e.g. eIF-6, HMG-CoA_red) tend to give these errors: _Error: the symbol 'B' accesing the matrix is not defined in this object_ or _Error: the symbol 'J' accesing the matrix is not defined in this object_. What could be causing this and does it have an effect on the subsequent steps/results? I have added a runlog to this message in which this can be seen.

Cheers and thanks
Guillaume
[nohup.txt](https://github.com/AstrobioMike/GToTree/files/8363266/nohup.txt)
e

**Comment by AstrobioMike** [comment]
Hey there, Guillame!

Thanks for the kind words :) 

> 1. I noticed that GToTree adds 5 Xs in the final alignment between the different aligned markers. Is there a reason for this?

It's common in concatenated alignments to put a few X's (with amino-acid alignments) as spacers between the individual alignments that are being stuck together. It's one of those things I did because it's convention and didn't give it too much thought before you brought it up, ha. I actually don't know if it's essential to have them, I suspect not, but those positions don't impact the tree estimation, and in the least they can make it easier for viewing and working with the concatenated alignment. 

> 2. When running the same input data (or different input data sets), I noticed that the order of the marker genes in Partitions.txt is not always the same. Sometimes it seems to be alphabetical, whereas sometimes it's more random (although the random one usually starts with the FbpA marker). Although this doesn't infuence the result, I was just wondering what influences this order.

Hmm, this is also something I hadn't looked into before – other than assuring they are done properly, regardless of the order of course. I tried 5 runs and happened to get the same order every time, but still looked into it a bit further. This is handled by this `gtt-cat-alignments` script [here](https://github.com/AstrobioMike/GToTree/blob/master/bin/gtt-cat-alignments), and specifically the order that happens for a given run is determined by a python `glob` function at this [line here](https://github.com/AstrobioMike/GToTree/blob/c70de4b8aa6c348973854911567c958ea8460312/bin/gtt-cat-alignments#L25) that grabs specific files in a directory (the individual alignment files in this case), and i don't know how `glob` is doing that. A response on [this stackoverflow](https://stackoverflow.com/a/6773661/10452709) suggests it might be system dependent. If it is pulling based on timestamp then they could vary for lots of reasons during the run just based on computational resources and how things are executed ¯\_(ツ)_/¯ 

> 3. During the Finished aligning and formatting gene-set step, some markers (e.g. eIF-6, HMG-CoA_red) tend to give these errors: Error: the symbol 'B' accesing the matrix is not defined in this object or Error: the symbol 'J' accesing the matrix is not defined in this object. What could be causing this and does it have an effect on the subsequent steps/results? I have added a runlog to this message in which this can be seen.

This is something coming from [trimAl](http://trimal.cgenomics.org/getting_started_with_trimal_v1.2), the automated trimming program used to hopefully help clean up the alignments prior to treeing. It's been a long time since i looked into how trimAl works, but in general I remember looking into these messages before, and I believe it happens when there are ambiguous residues in the amino acid alignment, in a column it is evaluating for whether or not it should be retained or trimmed (because it doesn't always give that message when there are just any ambiguous residues in there, if i remember correctly). The few times I've seen these error messages and looked, it still produced a reasonably trimmed alignment (e.g., still trimming ends and positions that have mostly gaps). I think maybe it can't optimally trim things because it doesn't have as much information on those columns holding these ambiguous residues, but it still does a reasonable job when I've looked and i don't think it's a concern overall if they show up on some

Thanks for the thoughtful questions! Hopefully my thoughts back are somewhat helpful :)
-Mike
